### ISSUE TITLE
Fix S1144 FP: Should not mark template-method implementations

### ISSUE THREAD

**Issue by Xilconic** [SonarSource/sonar-dotnet#9692]
### Description

S1144 is being raised unexpectedly for template-method implementations.

### Repro steps

```csharp
public class Repro()
{
    var x = new SomeClass();
    const string theName = x.Name;
    var y = x.DeepClone();
    const string theOtherName = y.Name;
}

public abstract class SomeClassDefiningTemplateMethods
{
    public abstract string Name { get; } // Template property
    public abstract SomeClassDefiningTemplateMethods DeepClone(); // Template method
}

public class SomeClass : SomeClassDefiningTemplateMethods
{
    public override string Name => "FooBar" // Currently being flagged with ``S1144: Remove the unused private property 'Name '.`
    public override SomeClassDefiningTemplateMethods DeepClone() // Currently being flagged with `S1144: Remove the unused private method 'DeepClone'.`
    {
        return new SomeClass();
    }
}
```

### Expected behavior

S1144 should not be raised by SonarAnalyzer.CSharp

### Actual behavior

For template method, raised: `S1144: Remove the unused private method 'DeepClone'.`
For template property, raises: `S1144: Remove the unused private property 'Distribution'.`
Also the message incorrectly states these are `private` methods/properties, while they actually are `public`.

### Known workarounds

Ignore S1144 with #pragma's around method/property overrides.

### Related information

* JetBrains Rider 2024.2.7
* dotnet version 8.0.303
* SonarAnalyzer.CSharp package v9.32.0.97167
* Windows 10

**Comment by Xilconic** [comment]
Seems related to https://github.com/SonarSource/sonar-dotnet/issues/4918, but I think that one was focussed on the raising items on `SomeClassDefiningTemplateMethods` from my example.

**Comment by Xilconic** [comment]
The false positive disappeared again at some point after building.
Maybe code caches of Rider are causing this?

Now that I think of it, the S1144 was raised in code, but not the build log.

**Comment by Xilconic** [comment]
Sorry, I took some shortcuts with writing the reproduction example from memory, based on what I was observing in my actual project.

I rewrote the provided example:
```csharp
public class Repro()
{
    public void SomeMethod()
    {
        var x = new SomeClass();
        string theName = x.Name;
        var y = x.DeepClone();
        string theOtherName = y.Name;
    }
}

public abstract class SomeClassDefiningTemplateMethods
{
    public abstract string Name { get; } // Template property
    public abstract SomeClassDefiningTemplateMethods DeepClone(); // Template method
}

public class SomeClass : SomeClassDefiningTemplateMethods
{
    public override string Name => "FooBar"; // Currently being flagged with `S1144: Remove the unused private property 'Name'.`
    public override SomeClassDefiningTemplateMethods DeepClone() // Currently being flagged with `S1144: Remove the unused private method 'DeepClone'.`
    {
        return new SomeClass();
    }
}
```

However it might not prove useful to you, as this particular sample does not raise the S1144 false positive I originally experienced in my original project 😞 
And as I mentioned in https://github.com/SonarSource/sonar-dotnet/issues/9692#issuecomment-2453434791, I don't have an actual reproduction scenario anymore in the original project either 😞 

Shall we close this issue? And in case I manage to reproduce it again, I'll create a new issue (or reopen this one)?
